### ISSUE TITLE
Webfont rendering on v3?

### ISSUE THREAD

**Issue by vincentwoo** [xtermjs/xterm.js#1164]
Playing with v3, it seems that on some browser hard refereshes, xterm doesn't seem to pick up the custom fontFamily I set and renders oddly:

![screenshot 2017-12-21 18 04 02](https://user-images.githubusercontent.com/613320/34282331-5de967bc-e679-11e7-918e-a79e328d06f7.png)

On a normal refresh after:

![screenshot 2017-12-21 18 04 35](https://user-images.githubusercontent.com/613320/34282340-6b04e282-e679-11e7-901d-4364954f7f80.png)

Is there a recommended flow for loading xterm with webfonts?

**Comment by mofux** [comment]
@vincentwoo Good point. Since we use the canvas for rendering, there is no automatic swap-out of fonts anymore. In general you have to make sure that the font is available when your terminal creation code runs. To do that, make sure the terminal creation code is loaded in head or body after the stylesheet that loads the font (see https://stackoverflow.com/questions/9899372/pure-javascript-equivalent-of-jquerys-ready-how-to-call-a-function-when-t). You could also force xterm.js into refreshing its char atlas by setting the fontFamily on the theme once you know that the font is available:

```javascript
term.setOption('theme', { fontFamily: 'YourFontFamily' })
```

@Tyriar Do you think we should handle this case automatically (refreshing the char atlas once the window finished loading)?

**Comment by vincentwoo** [comment]
It's tricky because even if your scripts occur at the end of `<body>` you're not guaranteed that all webfonts have loaded - the browser delays rendering until a character of a given font family is deemed needed. Ideally, xterm would use fallback fonts and refresh as necessary.

**Comment by Tyriar** [comment]
@vincentwoo I think you'll want to call open after the web fonts have finished loading, that would prevent flickering of a different font initially.

Anyone know if there an easy way to tell whether a font has been loaded/changed?

**Comment by vincentwoo** [comment]
It might be useful to have xterm incorporate one of the polyfills mentioned here: https://stackoverflow.com/questions/5680013/how-to-be-notified-once-a-web-font-has-loaded. Right now, anyone who wants to use a webfont with xterm will have to go through something similar, so it might have good utility for users as an addon. I can look into it if you don't have the time.

**Comment by Tyriar** [comment]
I'd prefer this to be done at a higher level than xterm.js, https://github.com/bramstein/fontfaceobserver is a lot of code to ship to everyone. As for making it an addon, I think we should eventually move away from the current addon model once we have a more complete API as currently the maintenance burden for all addons is put on the xterm.js team.

I recommend using fontfaceobserver on your side and only calling `Terminal.open` after they're loaded.

**Comment by vincentwoo** [comment]
Hm, even using fontfaceobserver to delay terminal creation, I sometimes see the wrong font loaded. Additionally, in Firefox, it appears to also cause some foreground text to render as black (on black). Has anyone spent a lot of time getting xterm 3.x working in production with webfonts?

**Comment by vincentwoo** [comment]
Here's a more detailed use case that I think warrants examination by the team. Let's take a pretty plausible situation: You're a user of xterm.js who wants to use a custom webfont for just terminal display. xterm is the only consumer of your special monospace webfont on your website. You might think, plausibly, that importing the proper font family, waiting an acceptable amount of time for the load, and then instantiating xterm would be enough.

This works well enough in Chrome. Firefox, instead of merely lowering the priority of fonts that are unused on the page at load time *actuals skips downloading those fonts altogether*. In this case, you get a black screen. See this codepen for an example: https://codepen.io/anon/pen/wmPJrQ.

In that pen, you'll see a `<span>` that you can uncomment which will force the browser to *actually* load the font. In xterm 2.x this was less of an issue because xterm would render to the DOM, and the presence of those DOM nodes with the correct `fontFamily` attribute would cause the browser to load the appropriate font. In the new world, *everything* needs to be correctly loaded before xterm can render. That includes, for instance, the bold variant of the font and any other separate variations.

I think there is a lot of virtue in xterm rendering a hidden dummy span which triggers the loading of any font variants it may need for the chosen `fontFamily` it has been configured with, and then deferring (or refreshing) when those fonts have been loaded.

**Comment by Tyriar** [comment]
I can see value in an addon that does this and then forces a repaint and refresh of the texture atlas after things are finished. Open to PRs.

**Comment by vincentwoo** [comment]
@Tyriar these internals are a bit hard for me. Is there a golden path way to force a refresh of the texture atlas?
